viewer/packages/tools/src/SmartOverlay/SmartOverlayTool.ts

```diff
@@ -0,0 +1,400 @@
+  }
+
+  set textOverlayVisible(visible: boolean) {
+    this._textOverlayVisible = visible;
+    this._textOverlay.style.opacity = visible ? '1' : '0';
+  }
+
+  get textOverlayVisible(): boolean {
+    return this._textOverlayVisible;
+  }
+
+  /**
+   * Removes all overlays.
+   */
+  clear(): void {
+    this._singeOverlayCollection.removeAllOverlays();
+
+    this._overlayPoints.forEach(points => {
+      this._viewer.removeObject3D(points);
+      points.dispose();
+    });
+    this._overlayPoints.splice(0, this._overlayPoints.length);
+
+    this._labelIndexToLabelInfoMap.clear();
+  }
+
+  on(event: 'hover', eventHandler: OverlayEventHandler): void;
+  on(event: 'click', eventHandler: OverlayEventHandler): void;
+
+  on(event: OverlayToolEvent, eventHandler: OverlayEventHandler): void {
+    switch (event) {
+      case 'hover':
+        this._events.hover.subscribe(eventHandler);
+        break;
+      case 'click':
+        this._events.click.subscribe(eventHandler);
+        break;
+      default:
+        assertNever(event);
+    }
+  }
+
+  off(event: 'hover', eventHandler: OverlayEventHandler): void;
+  off(event: 'click', eventHandler: OverlayEventHandler): void;
+
+  off(event: OverlayToolEvent, eventHandler: OverlayEventHandler): void {
+    switch (event) {
+      case 'hover':
+        this._events.hover.unsubscribe(eventHandler);
+        break;
+      case 'click':
+        this._events.click.unsubscribe(eventHandler);
+        break;
+      default:
+        assertNever(event);
+    }
+  }
+
+  private positionTextOverlay(event: MouseEvent): void {
+    const { _textOverlay, _textOverlayVisible } = this;
+    _textOverlay.style.left = `${event.offsetX}px`;
+    _textOverlay.style.top = `${event.offsetY}px`;
+    _textOverlay.style.opacity = _textOverlayVisible ? '1' : '0';
+  }
+
+  private intersectPointsMarkers(mouseCoords: {
+    clientX: number;
+    clientY: number;
+  }): { intersectedIcon: Overlay3DIcon; intersectedOverlay: OverlayInfo } | null {
+    const { _viewer, _raycaster, _temporaryVec } = this;
+
+    const { x, y } = this.convertPixelCoordinatesToNormalized(mouseCoords, _viewer.domElement);
+    const camera = _viewer.cameraManager.getCamera();
+    const cameraDirection = camera.getWorldDirection(new THREE.Vector3());
+    _raycaster.setFromCamera(_temporaryVec.set(x, y), camera);
+
+    const intersection: [Overlay3DIcon, THREE.Vector3][] = [];
+
+    for (const points of this._overlayPoints) {
+      for (const icon of points.icons) {
+        if (icon.visible) {
+          const inter = icon.intersect(_raycaster.ray);
+          if (inter) {
+            intersection.push([icon, inter]);
+            icon.updateAdaptiveScale({
+              camera,
+              renderSize: _viewer.renderParameters.renderSize,
+              domElement: _viewer.canvas
+            });
+          }
+        }
+      }
+    }
+
+    intersection.filter(([icon, _]) => icon.intersect(_raycaster.ray) !== null);
```

Comment:
This does nothing as you don't keep the results.

Reply:
Good catch, added rewriting of the result, works better now.